### PR TITLE
Friend widget 3

### PR DESCRIPTION
**Pull request by Diadlo**
This change is [Reviewable](https://reviewable.io/reviews/qtox/qtox/3881)

### CODE REVIEW

**Comment by sudden6** [comment]
:lgtm:

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881#-:-KWZYMpqrLqKhtlFySzv:bnfp4nl)_

**Comment by zetok** [comment]
on 00418ccfca636220e4fa35e9fb145b9d0ee074a0
- [x] #3886, should be green, is without color. To reproduce, have those steps done in a second groupchat; with only one groupchat there's no problem.
  spectacle ns5879

**Comment by zetok** [comment]
There are now some conflicts, rebase required.

**Comment by Diadlo** [comment]
@zetok Can you capture the video with bug?

**Comment by zetok** [comment]
@Diadlo 
[no-green-webm.zip](https://github.com/qTox/qTox/files/639800/no-green-webm.zip)

**Comment by Diadlo** [comment]
@zetok Fixed.

**Comment by zetok** [comment]
@Diadlo yeah, seems to be fixed.

**Comment by sudden6** [comment]
:lgtm:

---

Reviewed 3 of 10 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881#-:-K_9LTybX0Ft6wZMCLHv:bnfp4nl)*

**Comment by antis81** [comment]
Reviewed 4 of 10 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_9Qvgs12Jn4mne5hqG:b-lio7d9) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
> ```cpp
> 
>     friendList.insert(friendId, std::make_tuple(this, friendWidget));
>     emit friendWidget->chatroomWidgetClicked(friendWidget, false);
> ```

**Revert!**

Apart from the fact, that ContentDialog code is a complete bale of rubbish and need be replaced… yada blabla. Don't do such (toxic) things. Ask yourself the right questions - in this case "Was the (foreign) friend widget actually clicked?" IOW: The event is lying to the logic, which is a very bad thing actually and can leads to unpredictable code execution!

Basically the following rules apply:
1. A widget must not tell another widget about a (Tox) data change!
2. It is rarely required and far from good practice to emit a signal from another (foreign) object. A valid use case is from a C callback, that carries a `this` pointer to "self" - this is done at several places in Core for reference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by zetok** [comment]
PR needs to be rebased, since there are now some conflicts.

**Comment by Diadlo** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_BAmpenoS8fmyXLkaG:bc0d61m) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

**Revert!**

Apart from the fact, that ContentDialog code is a complete bale of rubbish and need be replaced… yada blabla. Don't do such (toxic) things. Ask yourself the right questions - in this case "Was the (foreign) friend widget actually clicked?" IOW: The event is lying to the logic, which is a very bad thing actually and can leads to unpredictable code execution!

Basically the following rules apply:
1. A widget must not tell another widget about a (Tox) data change!
2. It is rarely required and far from good practice to emit a signal from another (foreign) object. A valid use case is from a C callback, that carries a `this` pointer to "self" - this is done at several places in Core for reference.
</blockquote></details>

Just to make sure, that you understand the reason of this: we have 2 save friend widgets. And they should have same behaviour. 
1) If we will connect SIGNAL(clicked) to the few slots we should don't forget to call they here and in other place where not just `emit friendWidget->chatroomWidgetClicked`
2) If you criticize you should offer :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by Diadlo** [comment]
@antis81 I'm agree, that it's not the best solution. It's just lesser of two evils (this PR one in the sequence of PRs, which remove `FriendWidget` from `Friend`). I can add `TODO`

**Comment by antis81** [comment]
Reviewed 5 of 10 files at r4.
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/friend.cpp, line 41 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JVaEcFH2-4LeYxMLu:-K_JVaEd6JkbAusUs9uH:b-gvoym) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/friend.cpp#L41)):*
> ```cpp
> {
>     if (userName.isEmpty())
>         userName = UserId.getPublicKeyString();
> ```

revert brackets

---

*[src/persistence/offlinemsgengine.cpp, line 110 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JW1-ZxC14TMAL1zi8:-K_JW1-_MTGaawctlZbA:b-dbppvn) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/persistence/offlinemsgengine.cpp#L110)):*
> ```cpp
>         int rec;
>         if (val.msg->isAction())
>             rec = Core::getInstance()->sendAction(f->getFriendId(), messageText);
> ```

Restore the brackets.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_JYoRH8VQmIkIN5o6U:b-56cq7p) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Just to make sure, that you understand the reason of this: we have 2 save friend widgets. And they should have same behaviour. 
1) If we will connect SIGNAL(clicked) to the few slots we should don't forget to call they here and in other place where not just `emit friendWidget->chatroomWidgetClicked`
2) If you criticize you should offer :)
</blockquote></details>

Here's the offer:
Delete the line completely and **do not** activate the chatroom. This code is pure pile of toxic poop -- badly stinking one! :poop: :athletic_shoe: You won a ticket to the long journey of fixing broken concepts! Congrats pal! :smile:

---

*[src/widget/contentdialog.cpp, line 205 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JafoHaWAFRtDBBm2j:-K_JafoIZLbo6mpJExcN:brhutb5) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L205)):*
> ```cpp
> 
>     groupList.insert(groupId, std::make_tuple(this, groupWidget));
>     emit groupWidget->chatroomWidgetClicked(groupWidget, false);
> ```

Same here…

---

*[src/widget/contentdialog.cpp, line 360 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JajDqoJpSzfl9VLhq:-K_JajDrhhiOayhHU_ba:b8rjryy) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L360)):*
> ```cpp
> 
>         if (chatWidget != nullptr && chatWidget != activeChatroomWidget)
>             emit chatWidget->chatroomWidgetClicked(chatWidget, false);
> ```

…and here

---

*[src/widget/widget.cpp, line 1283 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JbwAJIv1rIt_5dZPS:-K_JbwAKXlv2iLAE94bs:bhb09o1) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1283)):*
> ```cpp
>         emit widget->chatroomWidgetClicked(widget, group);
>     });
>     emit widget->chatroomWidgetClicked(widget, false);
> ```

…and here (also the connect shit above)!

---

*[src/widget/widget.cpp, line 1324 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JcKXXfxafjTbo7OpK:-K_JcKXXfxafjTbo7OpL:b-ceeygk) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1324)):*
> ```cpp
>         emit widget->chatroomWidgetClicked(widget, group);
>     });
>     emit widget->chatroomWidgetClicked(widget, false);
> ```

…and here…

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by Diadlo** [comment]
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_JgvE13HxGNEdHfcu2:bvr3wrp) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

Here's the offer:
Delete the line completely and **do not** activate the chatroom. This code is pure pile of toxic poop -- badly stinking one! :poop: :athletic_shoe: You won a ticket to the long journey of fixing broken concepts! Congrats pal! :smile:
</blockquote></details>

It's breaking change. And it's out of the scope of current PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by Diadlo** [comment]
Review status: 5 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/friend.cpp, line 41 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JVaEcFH2-4LeYxMLu:-K_VqCa0A1gYIbJszCkU:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/friend.cpp#L41)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

revert brackets
</blockquote></details>

Done.

---

*[src/persistence/offlinemsgengine.cpp, line 110 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JW1-ZxC14TMAL1zi8:-K_VqD9FMY686XCw7P6j:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/persistence/offlinemsgengine.cpp#L110)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

Restore the brackets.
</blockquote></details>

Done.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_Vq2enCn0L4SXESOqV:br38zos) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

It's breaking change. And it's out of the scope of current PR
</blockquote></details>

@antis81 Do you agree to leave it now? (same for other `emit`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by antis81** [comment]
Reviewed 1 of 10 files at r4.
Review status: 6 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K__BYR6rwuw6TX90AhO:b-4mso5i) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

@antis81 Do you agree to leave it now? (same for other `emit`)
</blockquote></details>

Obviously I highly disagree with that! But… regarding the current workflow, which dictates "everything working all time" I don't see any chance to change anything. What can I say? Time is limited and I need to focus on more important stuff right now. Please bare with me and consider this as a "(temporary?) friendly exit" from qTox from my side. Looks like by upcoming weekend I might probably have something ready to show off - won't promise anything though, so don't rely on that please… :smile:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by Diadlo** [comment]
cc @sudden6

---

Review status: 6 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K__KlMvVuZm7DDhJn09:bba7h0q) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

Obviously I highly disagree with that! But… regarding the current workflow, which dictates "everything working all time" I don't see any chance to change anything. What can I say? Time is limited and I need to focus on more important stuff right now. Please bare with me and consider this as a "(temporary?) friendly exit" from qTox from my side. Looks like by upcoming weekend I might probably have something ready to show off - won't promise anything though, so don't rely on that please… :smile:
</blockquote></details>

Thank you entered my position. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881#-:-K__L7pq3Sn2CQKizqVk:b-b8mo4a)*

**Comment by sudden6** [comment]
Reviewed 2 of 2 files at r5.
Review status: 8 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K__M2RhUxiBHnTor3c6:bv1z037) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Thank you entered my position. 
</blockquote></details>

make a `FIXME:` here so we remember it's a hack

---

*[src/widget/contentdialog.cpp, line 205 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JafoHaWAFRtDBBm2j:-K__MC7SSCQahVBGL-rj:b-cjuhw1) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L205)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

Same here…
</blockquote></details>

also add a `FIXME` here

---

*[src/widget/contentdialog.cpp, line 360 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JajDqoJpSzfl9VLhq:-K__MI42Fj-ZkcpCp-S9:b-cjuhw1) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L360)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

…and here
</blockquote></details>

also add a `FIXME` here

---

*[src/widget/widget.cpp, line 1283 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JbwAJIv1rIt_5dZPS:-K__N2p7pn5PEzSHXlko:b-cjuhw1) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1283)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

…and here (also the connect shit above)!
</blockquote></details>

also add a `FIXME` here

---

*[src/widget/widget.cpp, line 1324 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JcKXXfxafjTbo7OpK:-K__N4EttDZv7XVSiEsO:b-cjuhw1) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1324)):*
<details><summary><i>Previously, antis81 (Nils Fenner) wrote…</i></summary><blockquote>

…and here…
</blockquote></details>

also add a `FIXME` here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by Diadlo** [comment]
Review status: 8 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/widget/contentdialog.cpp, line 184 at r3](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_9QvgrAKp9kqhANH-r:-K_adjekOdYetnfUe5Mz:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/8e45e55d567daa91fa7057d5adba9b4c15fb1367/src/widget/contentdialog.cpp#L184)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

make a `FIXME:` here so we remember it's a hack
</blockquote></details>

Done.

---

*[src/widget/contentdialog.cpp, line 205 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JafoHaWAFRtDBBm2j:-K_adkF8-88ipywCIA05:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L205)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

also add a `FIXME` here
</blockquote></details>

Done.

---

*[src/widget/contentdialog.cpp, line 360 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JajDqoJpSzfl9VLhq:-K_adkwmGJqyks0HrA-H:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/contentdialog.cpp#L360)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

also add a `FIXME` here
</blockquote></details>

Done.

---

*[src/widget/widget.cpp, line 1283 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JbwAJIv1rIt_5dZPS:-K_adlUaNi8eercBI_R8:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1283)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

also add a `FIXME` here
</blockquote></details>

Done.

---

*[src/widget/widget.cpp, line 1324 at r4](https://reviewable.io:443/reviews/qtox/qtox/3881#-K_JcKXXfxafjTbo7OpK:-K_adm86osrXFBoPip5w:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/006dc90d568fed7099fab7e05f74f256c930c8be/src/widget/widget.cpp#L1324)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

also add a `FIXME` here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881)*

**Comment by sudden6** [comment]
:lgtm_strong: 

---

Reviewed 6 of 10 files at r4, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/3881#-:-K_b-5egGjoZMzQ1fTK7:b-uwkowd)*
